vignettes/opencl.Rmd

```diff
@@ -0,0 +1,143 @@
+---
+title: "Run Stan on the GPU with OpenCL"
+author: "Rok Češnovar and Jonah Gabry"
+output: 
+  rmarkdown::html_vignette:
+    toc: true
+    toc_depth: 4
+params:
+  EVAL: !r identical(Sys.getenv("CMDSTANR_OPENCL_TESTS"), "true")
+vignette: >
+  %\VignetteIndexEntry{Run Stan on the GPU with OpenCL}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## Introduction
+
+This vignette demonstrates how to use the OpenCL capabilites of CmdStan with
+CmdStanR. The vignette requires CmdStan 2.26.1 or newer.
+
+As of version 2.26.1, users can expect speedups with OpenCL when using vectorized
+probability distribution/mass functions (functions with the `_lpdf` or `_lpmf`
+suffix). You can expect speedups when the input variables contain 20.000 or more elements.
+
+The actual speedup for a model will depend on whether the `lpdf/lpmf` functions
+are the bottlenecks of the model and the `lpdf/lpmf` function used. 
+The more computationally complex the function is, the larger the expected speedup.
+The biggest speedups are expected when using the GLM functions.
```

Comment:
Man this is great!

Reply:
Just one clarification, what are 20.000 elements? Is that data or input parameters? I just want to make sure that the vignette has no ambiguities.

Reply:
The 20k elements here are meant as N of the input vectors/row_vectors/arrays for the worst-case.

If most of them are data, the speedups will be bigger and typically a GPU will be faster than a CPU at smaller N.
The worst-case is all of the inputs are vector parameters as there is a lot more copies to be done there.

See for example:
![image](https://user-images.githubusercontent.com/28476796/113401016-06533200-93a3-11eb-886f-595e29660f8f.png)

The best case is the non-data inputs are scalars:

![image](https://user-images.githubusercontent.com/28476796/113401067-1a972f00-93a3-11eb-8f92-e6cdf45a6f18.png)

The theoretical best-case is all inputs are data, but an all-data lpdf/lpmf call in the model/transformed parameters blocks are just inefficient anyways. Its still shown on the above charts as that will be achievable once we turn on support for non-lpdf/lpmf functions as well (the inputs are calculated on the GPU -> less copies).

This is a "ballpark" number, as mentioned it will depend on the exact CPU/GPU/lpdf/lpmf.

Reply:
Thank you!